packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -923,8 +927,13 @@ class HeroController extends NavigatorObserver {
 
     // For pop transitions driven by a user gesture: if the "to" page has
     // maintainState = true, then the hero's final dimensions can be measured
-    // immediately because their page's layout is still valid.
-    if (isUserGestureTransition && flightType == HeroFlightDirection.pop && toRoute.maintainState) {
+    // immediately because their page's layout is still valid. Unless due to directly
+    // adding routes to the pages stack causing the route to never get laid out.
+    final bool needsLayout = toRoute.subtreeContext?.findRenderObject()?.debugNeedsLayout ?? true;
```

Comment:
@chunhtai do you happen to know if this check would still work in prod apps, or if there's a better way to check for if a route has done it's layout logic yet?

Reply:
no the getter is wrapped in an assert and will crash in release mode, and any method with debug keyword should only be called in assert anyway

Reply:
As far as I know there is no direct way to know whether a renderobject is layout dirty or not. Can you do something like RenderBox.hasSize?